docs/indexing.md

```diff
@@ -0,0 +1,422 @@
+
+### Bitcast
+
+A bitcast op can be represented as a [sequence of transpose-reshape-transpose](https://github.com/openxla/xla/blob/578b6df240be94c3c84129fd83f34487efc623a5/xla/shape_util.h#L813). Therefore, its indexing maps are just a composition of indexing maps for this sequence.
+
+### Concatenate
+
+Output-to-input mapping for concat is defined for all inputs, but with non-overlapping domains, i.e. only one of the inputs will be used at a time.
+
+```c+
+p0 = f32[3,50] parameter(0)
+p1 = f32[3,30] parameter(1)
+concat = f32[3,80] concatenate(f32[3,50] p0, f32[3,30] p1),
+  dimensions={1}
+```
+
+The output to input map:
+
+- output -> input 1:
+
+$(d_0, d_1) \mapsto (d_0, d_1)$ for $\boldsymbol{d} \in [0, 2] \times [0, 49]$
+
+- output -> input 2:
+
+$(d_0, d_1) \mapsto (d_0, d_1 - 50)$ for $\boldsymbol{d} \in [0, 2] \times [50, 79]$
+
+The inputs to output map:
+
+- input 1 -> output: $(d_0, d_1) \mapsto (d_0, d_1)$ for $\boldsymbol{d} \in {\rm Dom}(input_1)$.
+- input 2 -> output: $(d_0, d_1) \mapsto (d_0, d_1 + 50)$ for $\boldsymbol{d} \in {\rm Dom}(input_2)$.
+
+### Dot (output-to-input implemented
+
+Indexing maps for dot are very similar to the ones of reduce.
+
+```c+
+p0 = f32[4, 128, 256] parameter(0)
+p1 = f32[4, 256, 64] parameter(1)
+dot = f32[4, 128, 64] dot(p0, p1), 
+  lhs_batch_dims={0}, rhs_batch_dims={0},
+  lhs_contracting_dims={2}, rhs_contracting_dims={1}
+```
+The output to inputs maps:
+
+- output -> input_1: $(d_0, d_1, d_2) \mapsto (d_0, d_1, s_0)$ for $\boldsymbol{d} \in {\rm Dom}(output)$ and $\boldsymbol{s} \in [0, 255]$
+- output -> input_2: $(d_0, d_1, d_2) \mapsto (d_0, s_0, d_2)$ for $\boldsymbol{d} \in {\rm Dom}(output)$ and $\boldsymbol{s} \in [0, 255]$
+
+The inputs to output maps:
+
+- input_1 -> output: $(d_0, d_1, d_2) \mapsto (d_0, d_1, s_0)$ for $\boldsymbol{d} \in {\rm Dom}(input_1)$ and $\boldsymbol{s} \in [0, 63]$
+- input_2 -> output: $(d_0, d_1, d_2) \mapsto (d_0, s_0, d_1)$ for $\boldsymbol{d} \in {\rm Dom}(input_2)$ and $\boldsymbol{s} \in [0, 127]$
+
+### Reduce-window (TBD)
+
+### Pad (TBD)
+
+
+
+## Indexing Maps for Fusion
+
+Indexing map for fusion op is a composition of indexing maps for every op in the cluster. It can happen that some inputs are read several times with different access patterns.
+
+### One input, several indexing maps
+
+Here is an example for $p_0 + p_0^T$
+
+```c+
+f {  
+  p0 = f32[1000, 1000] parameter(0)
+  transpose_p0 = f32[1000, 1000]{0, 1}
+  transpose(p0), dimensions={1, 0}
+  ROOT a0 = f32[1000, 1000] add(p0, transpose_p0)
+}
+```
+
+The output-to-input indexing maps for `p0` will be $(d_0, d_1) \mapsto (d_0, d_1)$ and $(d_0, d_1) \mapsto (d_1, d_0)$.
+
+### One input, deduplicated indexing map
+
+![img](./images/indexing_analysis_transposes.svg)
+
+There are cases when the indexing maps are actually the same, even though it is not immediately obvious. 
```

Comment:
So we could replace four `transpose` ops with a single one in HLO, right?

Reply:
Yes, if we swap transpose and exponential.